### PR TITLE
New Props API

### PR DESCRIPTION
**Pull request by natemoo-re**
## Changes

Astro's parser was originally forked from the Svelte parser, so we inherited their syntax for component props. This PR updates to a new, Astro-specific props syntax.

See #139, #309

**Previously**
```jsx
---
export let title: string | undefined;
---
```

**New Props API**
```jsx
export interface Props {
  title?: string;
}
const { title } = Astro.props as Props;
```

## Testing

TBD

## Docs

TBD

### CODE REVIEW

**Comment by matthewp** [comment]
@natemoo-re Code looks pretty good to me, need any help pushing this along?

**Comment by natemoo-re** [comment]
@matthewp Just need to go through and update our examples! I'm doing that now.

File: docs-www/src/layouts/Main.astro [hunk 1]
```diff
@@ -4,7 +4,7 @@ import SiteSidebar from '../components/SiteSidebar.astro';
 import ThemeToggle from '../components/ThemeToggle.tsx';
 import DocSidebar from '../components/DocSidebar.tsx';
 
-export let content;
+const { content } = Astro.props;
```
Comment (matthewp) [inline docs-www/src/layouts/Main.astro:7]: i like this so much better, no magic.

**Review by matthewp — CHANGES_REQUESTED** [review]
Code looks good. Can you add a changeset and make this a minor change? This will break existing apps so want to make sure it doesn't go through as a patch on accident.

**Comment by FredKSchott** [comment]
This looks great! Can we add a warning if you try to define props the old way? it should be as easy as scanning the exports, and only allowing `createCollection` and `Props`. And then we can remove this workaround whenever we tackle the idea of a component export.

**Review by matthewp — APPROVED** [review]
Great changeset!

**Comment by natemoo-re** [comment]
@FredKSchott warning added!

Screen Shot 2021-06-24 at 5 33 59 PM

File: packages/astro/src/compiler/codegen/index.ts [hunk 2]
```diff
@@ -411,12 +409,16 @@ function compileModule(module: Script, state: CodegenState, compileOptions: Comp
...
       for (const importStatement of globResult.imports) {
-        state.importExportStatements.add(importStatement);
+        state.importStatements.add(importStatement);
       }
       contentCode += globResult.code;
     }
 
     script = propsStatement + contentCode + babelGenerator(program).code;
+    const location = { start: module.start, end: module.end };
+    let transpiledScript = compileExpressionSafe(script, { state, compileOptions, location });
```
Comment (duncanhealy) [inline packages/astro/src/compiler/codegen/index.ts:419]:
```suggestion
 let transpiledScript = transpileExpressionSafe(script, { state, compileOptions, location });
```

**Review by duncanhealy — COMMENTED** [review]
compileExpressionSafe got changed by pr https://github.com/snowpackjs/astro/pull/522  to transpileExpressionSafe

File: packages/astro/src/compiler/codegen/index.ts (hunk 2 above)
Comment (jasikpark) [reply packages/astro/src/compiler/codegen/index.ts:419]: Yup. This is failing to build for me D:

**Pull request merged**